### ISSUE TITLE
error MSB4181 when using NuGet CPVM

### ISSUE THREAD

**Issue by EugeneKrapivin** [kaby76/Antlr4BuildTasks#20]
I'm attempting to move my solution to [NuGet CPVM (Central Package Version Management)](https://github.com/NuGet/Home/wiki/Centrally-managing-NuGet-package-versions), but alas after the move I'm getting the following error while building:

```
Antlr4BuildTasks.targets(195,3): error MSB4181: The "RunAntlrTool" task returned false but did not log an error.
```

could be seen at this [build flow](https://github.com/EugeneKrapivin/JsonTranslate.NET/runs/3483131835?check_suite_focus=true#step:5:28)
specifically build fails on [this branch](https://github.com/EugeneKrapivin/JsonTranslate.NET/tree/moving-to-nuget-cpvm)

Any ideas why would that happen?

**Comment by kaby76** [comment]
The easiest way to fix the bug is to change the build temporarily to "dotnet build --no-restore -v diag" then examine the output.

**Comment by kaby76** [comment]
Right. Harwell's old code sets trace level to "info" for exceptions. Doesn't make sense, since if there's a crash, *let us know why!*. There are a number of bugs I still haven't cleaned out of Harwell's old Antlr4 build code. I will release a 8.15 version asap, then you can try that, if you haven't tried "-v diag" earlier.

**Comment by kaby76** [comment]
Eugene, If you have time, please try Antlr4BuildTasks 8.15 and see what it says. The error reporting is now fixed, but I don't know why the build would fail. Thanks.

**Comment by kaby76** [comment]
OK. The way the Antlr4BuildTasks tool works is to:

1) Find the Java JRE.
2) Find the Antlr4 Tool Jar.
3) Call the Antlr4 Tool Jar using the Java JRE and munch on your Antlr4 grammars.
4) Manage what files are generated, what needs to be added to the build, what needs to be removed on a Clean.

For your build, it finds the Java Runtime no problem. That's because it's included in the package and returns that. The tool has some environmental variables or msbuild variables it can look at, but you don't set any of those, so it returns the default.

Then GetAntlrJar task is called. That's supposed to figure out where the Antlr Tool jar is. GetAntlrJar returns an empty string, which means it couldn't find the Antlr Tool jar.

So, the first problem is that if it's going to return null or an empty string, then it needs to throw an error. It should not play dumb. And it is.

The AntlrToolJar task doesn't say what it's looking for, which makes it hard for me looking at the output what *version of the Antlr4.Runtime.Standard* it's probing for. Also a dumb oversight on my part.

But, your .csproj is very different. It doesn't contain a version for Antlr4.Runtime.Standard in the .csproj. It's instead in a [props file](https://github.com/EugeneKrapivin/JsonTranslate.NET/blob/moving-to-nuget-cpvm/Directory.Packages.props#L5) way, way over in left field. GetAntlrJar is passed `@PackageReference`, with Antlr4.Runtime.Standard but without a version. So, [this line](https://github.com/kaby76/Antlr4BuildTasks/blob/master/Antlr4BuildTasks/GetAntlrJar.cs#L67) likely fails, and so it has no idea what to find.

**The solution for now is to just add `version="4.9.2"` attribute in the `PackageReference` element.**

In the meanwhile, I'll try to figure out how to get the version it's looking for in this new fangled "Centrally managing NuGet package versions" way of build files where it's placed way outside the .csproj file.

**Comment by kaby76** [comment]
Fix in version 8.16. The solution involves passing a "PackageVersion" parameter to the GetAntlrJar task. The parameter is the collection of all PackageVersion items, wherever they be, in this case in the file Directory.Packages.props.  If the version is set in the PackageReference element in the .csproj, it will take that. Otherwise, it searches through PackageVersion's and finds the version on that meta. If that is empty, then the GetAntlrJar task fails and prints out both lists.

**Comment by EugeneKrapivin** [comment]
I've tested it locally and it looks awesome, thanks for the kind and swift response!
